src/plot_api/subroutines.js

```diff
@@ -103,25 +116,21 @@ function lsInner(gd) {
         ax._mainSubplot = findMainSubplot(ax, fullLayout);
     }
 
-    fullLayout._paperdiv
-        .style({
-            width: fullLayout.width + 'px',
-            height: fullLayout.height + 'px'
-        })
-        .selectAll('.main-svg')
-            .call(Drawing.setSize, fullLayout.width, fullLayout.height);
-
-    gd._context.setBackground(gd, fullLayout.paper_bgcolor);
-
-    var subplotSelection = fullLayout._paper.selectAll('g.subplot');
-
-    // figure out which backgrounds we need to draw, and in which layers
-    // to put them
+    // figure out which backgrounds we need to draw,
+    // and in which layers to put them
     var lowerBackgroundIDs = [];
+    var backgroundIds = [];
     var lowerDomains = [];
-    subplotSelection.each(function(d) {
-        var subplot = d[0];
-        var plotinfo = fullLayout._plots[subplot];
+    // no need to draw background when paper and plot color are the same color,
+    // activate mode just for large splom (which benefit the most from this
+    // optimization), but this could apply to all cartesian subplots.
+    var noNeedForBg = (
+        fullLayout._hasOnlyLargeSploms &&
+        fullLayout.paper_bgcolor === fullLayout.plot_bgcolor
```

Comment:
oh interesting - nice find! I guess this is what you're implying, but would `paper_bgcolor === plot_bgcolor` be sufficient? Is there something special about `splom` in this regard?

Also I guess to be precise we need to know that the color is fully opaque too (or fully transparent)... though a semitransparent `paper_bgcolor` is a strange and dangerous thing.

`overlappingDomain` might itself be a heavy lift for a large number of subplots (am I reading this right that for NxN subplots this is O(N^4), ie O(N^2) tests per subplot???), all for the fairly unusual case of inset plots, or otherwise partially occluding subplots. Perhaps it would be worthwhile doing a faster check for whether `overlappingDomain` is even necessary to call, by looping over all x axes to see if any of their domains overlap, then the same over all y axes? That would only be O(N^2) to possibly (usually!) avoid O(N^4).

Reply:
> Is there something special about `splom` in this regard?

There's nothing special about splom in this regard. I added a check for  `_hasOnlyLargeSploms` for two (somewhat poor) reasons:

- I was scared of potentially missing a funky inset/subplot scenario
- many jasmine tests were failing as the custom assertion `assertPlotSize` relies on the [`<rect.bg>` bounding box](https://github.com/plotly/plotly.js/blob/5c62c8e937c92f3c267722b29613ebe2a15413c2/test/jasmine/assets/custom_assertions.js#L264) and I couldn't find a good replacement for it.

> Also I guess to be precise we need to know that the color is fully opaque too (or fully transparent)... though a semitransparent paper_bgcolor is a strange and dangerous thing.

Good call. I'll add a check for this.

> overlappingDomain might itself be a heavy lift for a large number of subplots (am I reading this right that for NxN subplots this is O(N^4), ie O(N^2) tests per subplot???), all for the fairly unusual case of inset plots, or otherwise partially occluding subplots. Perhaps it would be worthwhile doing a faster check for whether overlappingDomain is even necessary to call, by looping over all x axes to see if any of their domains overlap, then the same over all y axes? That would only be O(N^2) to possibly (usually!) avoid O(N^4).

That loop clocks in at 20-40ms at 50 dimensions, so I didn't bother optimizing it for now. You're right, it is an expansive loop, but it doesn't query nor append anything on the DOM, so it's not too too slow.

But yeah, it might be worth experimenting restricting the `hasOnlyLargeSploms` regime to graphs with only splom-generated cartesian subplots (i.e. no insets, no overlaps). This may help us speed up `drawFramework` too.

Reply:
OK, other than the semitransparent check these are nonblocking for this PR but please put removing the `_hasOnlyLargeSploms` check and avoiding the O(N^4) loop into issues somewhere so we don't forget about them.

Reply:
>  other than the semitransparent check these are nonblocking for this PR but please put removing the _hasOnlyLargeSploms check 

Commit https://github.com/plotly/plotly.js/pull/3057/commits/aaf63127bbad48adef96a19a4d456b7e66ce68b5 implements this optimization to all cartesian subplots and add a check for semi-transparent bg colors.

> avoiding the O(N^4) loop into issues somewhere so we don't forget about them.

Now in https://github.com/plotly/plotly.js/issues/2555#issuecomment-426343601